zkevm-circuits/src/test_util.rs

```diff
@@ -42,130 +41,201 @@ impl Default for BytecodeTestConfig {
-    test_ctx: TestContext<NACC, NTX>,
-    config: Option<BytecodeTestConfig>,
-) -> Result<(), Vec<VerifyFailure>> {
-    let block: GethData = test_ctx.into();
-    let mut builder =
-        BlockData::new_from_geth_data_with_params(block.clone(), CircuitsParams::default())
-            .new_circuit_input_builder();
-    builder
-        .handle_block(&block.eth_block, &block.geth_traces)
-        .unwrap();
+/// Struct used to easily generate tests for EVM &| State circuits being able to
+/// customize all of the steps involved in the testing itself.
+///
+/// By default, the tests run through `prover.assert_satisfied_par()` but the
+/// builder pattern provides functions that allow to pass different functions
+/// that the prover should execute when verifying the CTB correctness.
+///
+/// The CTB also includes a mechanism to recieve calls that will modify the
+/// block produced from the [`TestContext`] and apply them before starting to
+/// compute the proof.
+///
+/// ## Example:
+/// ```rust, no_run
+/// use eth_types::geth_types::Account;
+/// use eth_types::{address, bytecode, Address, Bytecode, ToWord, Word, U256, word};
+/// use mock::{TestContext, MOCK_ACCOUNTS, gwei, eth};
+/// use zkevm_circuits::test_util::CircuitTestBuilder;
+///     let code = bytecode! {
+/// // [ADDRESS, STOP]
+///     PUSH32(word!("
+/// 3000000000000000000000000000000000000000000000000000000000000000"))
+///     PUSH1(0)
+///     MSTORE
+///
+///     PUSH1(2)
+///     PUSH1(0)
+///     RETURN
+/// };
+/// let ctx = TestContext::<1, 1>::new(
+///     None,
+///     |accs| {
+///         accs[0].address(MOCK_ACCOUNTS[0]).balance(eth(20));
+///     },
+///     |mut txs, _accs| {
+///         txs[0]
+///             .from(MOCK_ACCOUNTS[0])
+///             .gas_price(gwei(2))
+///             .gas(Word::from(0x10000))
+///             .value(eth(2))
+///             .input(code.into());
+///     },
+///     |block, _tx| block.number(0xcafeu64),
+/// )
+/// .unwrap();
+///
+/// CircuitTestBuilder::empty()
+///     .test_ctx(ctx)
+///     .block_modifier(Box::new(|block| block.evm_circuit_pad_to = (1 << 18) - 100))
+///     .state_checks(Box::new(|prover| assert!(prover.verify_par().is_err())))
+///     .run();
+/// ```
+pub struct CircuitTestBuilder<const NACC: usize, const NTX: usize> {
+    test_ctx: Option<TestContext<NACC, NTX>>,
+    bytecode_config: Option<BytecodeTestConfig>,
+    circuit_params: Option<CircuitsParams>,
+    block: Option<Block<Fr>>,
+    evm_checks: Option<Box<dyn Fn(MockProver<Fr>)>>,
+    state_checks: Option<Box<dyn Fn(MockProver<Fr>)>>,
+    block_modifiers: Vec<Box<dyn Fn(&mut Block<Fr>)>>,
+}
 
-    // build a witness block from trace result
-    let block = crate::witness::block_convert(&builder.block, &builder.code_db).unwrap();
+impl<const NACC: usize, const NTX: usize> CircuitTestBuilder<NACC, NTX> {
+    /// Generates an empty/set to default `CircuitTestBuilder`.
+    pub fn empty() -> Self {
```

Comment:
This is a nitpick: I see that an "empty" CircuitTestBuilder is not usable.  Either it has a block or a test_ctx.  So what about instead of having an `empty` constructor it has:
- `new_from_block`
- `new_from_test_ctx`
Or something like that?

Reply:
Looks good to me!
Added in 3013511b